### PR TITLE
Implement kpod rm

### PR DESCRIPTION
**Pull request by 14rcole**
Kpod rm removes a container from the system

### CODE REVIEW

File: cmd/kpod/rm.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	config, err := getConfig(c)
+	if err != nil {
+		return errors.Wrapf(err, "could not get config")
+	}
+	server, err := libkpod.New(config)
+	if err != nil {
+		return errors.Wrapf("could not get container server")
+	}
+
+	fmt.Println(libkpod.RemoveImages(c.Args(), c.Bool("force")))
```
Comment (rhatdan) [inline cmd/kpod/rm.go]: Shouldn't this be RemoveContainers?

File: libkpod/remove.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+func (c *ContainerServer) remove(container string, force bool) (string, error) {
+	ctr, err := c.LookupContainer(container)
+	if err != nil {
+		return "", err
+	}
+	ctrID := ctr.ID()
+
+	if err := c.runtime.UpdateStatus(ctr); err != nil {
+		return ctrID, fmt.Errorf("failed to update container state: %v", err)
```
Comment (rhatdan) [inline libkpod/remove.go]: We usually return "" on errors.

File: libkpod/remove.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	cState := c.runtime.ContainerStatus(ctr)
+	if cState.Status == oci.ContainerStateCreated || cState.Status == oci.ContainerStateRunning {
+		if err := c.runtime.StopContainer(ctr, -1); err != nil {
+			return ctrID, fmt.Errorf("failed to stop container %s: %v", ctrID, err)
```
Comment (rhatdan) [inline libkpod/remove.go]: We usually return "" on errors.

File: libkpod/remove.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		}
+		if err := c.storageRuntimeServer.StopContainer(ctrID; err != nil {
+			return ctrID, fmt.Errorf("failed to unmount container %s: %v", ctrID, err)
```
Comment (rhatdan) [inline libkpod/remove.go]: We usually return "" on errors.

File: libkpod/remove.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+// RemoveContainers removes the given containers. If the container is running, the container
+// should be force removed.
+func (c *ContainerServer) RemoveContainers(containers []string, force bool) string {
+	var ret string
+	for _, container := range containers {
+		id, err := c.remove(container, force)
+		if err != nil {
+			ret = append(ret, err.Error(), "\n")
```
Comment (rhatdan) [inline libkpod/remove.go]: Should we return a list of containers that failed and the errors that were returned?

**Comment by rhatdan** [comment]
Missing man page, link to main.go, bash completions.

File: libkpod/remove.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"fmt"
+
+	"github.com/kubernetes-incubator/cri-o/oci"
+	"github.com/sirupsen/logrus"
+)
+
+// RemoveContainers removes the given containers. If the container is running, the container
+// should be force removed.
+func (c *ContainerServer) RemoveContainers(containers []string, force bool) string {
```
Comment (mheon) [inline libkpod/remove.go]: Why make this endpoint remove a list of containers? Docker's remove container API endpoint targets a single container, and then it implements handling for multiple containers in the CLI. I think that's the right model here - keeps the CLI simple.

File: libkpod/remove.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	ctrID := ctr.ID()
+
+	if err := c.runtime.UpdateStatus(ctr); err != nil {
+		return ctrID, fmt.Errorf("failed to update container state: %v", err)
+	}
+
+	cState := c.runtime.ContainerStatus(ctr)
+	if cState.Status == oci.ContainerStateCreated || cState.Status == oci.ContainerStateRunning {
+		if err := c.runtime.StopContainer(ctr, -1); err != nil {
```
Comment (mheon) [inline libkpod/remove.go]: We shouldn't stop the container in this API. If --force is passed to the CLI, we should stop the container there.

File: libkpod/remove.go (hunk 6 above)
Comment (rhatdan) [inline libkpod/remove.go]: I agree, lets just export RemoveContainer

File: libkpod/remove.go (hunk 7 above)
Comment (mheon) [inline libkpod/remove.go]: We should also check that the container is stopped when we remove it.

Comment (14rcole) [inline libkpod/remove.go]: I've been trying to put all of the logic into libkpod and basically leave kpod as an I/O layer.  By checking if a container is stopped, we have to look up the container and get the state within the cli, all of which seems like logic that should be in libkpod for others to use if the want to.  I did change the logic to check if the container was stopped though

File: libkpod/remove.go (hunk 5 above)
Comment (14rcole) [inline libkpod/remove.go]: Docker prints out either the container id or an error message for each container.  I wasn't sure how to handle that with the way we report errors through the cli package.  Take a look at what I changed it to in my most recent push and see if that's any better

File: libkpod/remove.go (hunk 7 above)
Comment (mheon) [inline libkpod/remove.go]: From an API design standpoint, I don't think we should have any one endpoint do multiple things. Remove should remove, not stop, by default. This keeps the API easy to understand and simple to implement.

Comment (mheon) [inline libkpod/remove.go]: On kpod being an I/O layer: I don't think that's the direction we're going. Libkpod is not just a backend for CRI-O and kpod. We should design things from a way that makes sense for other code that would want to use libkpod to interact with containers.

Comment (mheon) [inline libkpod/remove.go]: In short, I see no issue with putting logic into the kpod binary itself if it keeps our API clean and sensible.

File: libkpod/remove.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
+package libkpod
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-incubator/cri-o/oci"
+)
+
+// Remove removes a r
+func (c *ContainerServer) Remove(container string, force bool) (string, error) {
```
Comment (mheon) [inline libkpod/remove.go]: Can we change Server to call into these, and remove the duplicate implementations there? The semantics will be slightly different, but shouldn't be that hard to match.

File: libkpod/stop.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"fmt"
+
+	"github.com/kubernetes-incubator/cri-o/oci"
+	"github.com/sirupsen/logrus"
+	pb "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+)
+
+// Stop stops a running container with a grace period (i.e., timeout).
+func (c *ContainerServer) Stop(container string, timeout int64) (string, error) {
```
Comment (mheon) [inline libkpod/stop.go]: See above - can we remove the duplicate implementation in Server and call into this as well?

File: libkpod/remove.go (hunk 8 above)
Comment (mheon) [inline libkpod/remove.go]: Also, can we name this ContainerRemove? We're eventually going to have a PodRemove, too

File: libkpod/stop.go (hunk 9 above)
Comment (mheon) [inline libkpod/stop.go]: As above, please rename to ContainerStop

File: completions/bash/kpod [hunk 10]
```diff
@@ -6,6 +6,10 @@ __kpod_list_images() {
     COMPREPLY=($(compgen -W "$(kpod images -q)" -- $cur))
 }
 
+__kpod_list_containers() {
+    COMPREPLY=($(compgen -W "$(kpod ps -aq)" -- $cur))
```
Comment (rhatdan) [inline completions/bash/kpod]: Do we have a kpod ps?

File: completions/bash/kpod [hunk 11]
```diff
@@ -96,6 +100,25 @@ _kpod_push() {
...
+
+    local options_with_args="
+    "
+
+    case "$cur" in
+        -*)
+            COMPREPLY=($(compgen -W "$boolean_options $options_with_args" -- "$cur"))
+            ;;
+        *)
+            __kpod_list_images
```
Comment (rhatdan) [inline completions/bash/kpod]: kpod_list_containers?

File: docs/kpod-rm.1.md [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+
+## EXAMPLE
+
+kpod rm mywebserver
+
+kpod rm -f 860a4b23
+
+## SEE ALSO
+kpod(1)
```
Comment (rhatdan) [inline docs/kpod-rm.1.md]: Add kpod-rmi(1) since you mentioned removing images.

File: libkpod/remove.go [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+// Remove removes a r
+func (c *ContainerServer) ContainerRemove(container string, force bool) (string, error) {
+	ctr, err := c.LookupContainer(container)
+	if err != nil {
+		return "", err
+	}
+	ctrID := ctr.ID()
+
+	if err := c.runtime.UpdateStatus(ctr); err != nil {
+		return "", fmt.Errorf("failed to update container state: %v", err)
```
Comment (rhatdan) [inline libkpod/remove.go]: errors.Wrapf(err, "Failed to update container %v state", ctrID)

File: libkpod/remove.go [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	cState := c.runtime.ContainerStatus(ctr)
+	if cState.Status == oci.ContainerStateRunning {
+		if force {
+			_, err = c.ContainerStop(container, -1)
+			if err != nil {
+				return "", fmt.Errorf("unable to stop container %s: %v", ctrID, err)
```
Comment (rhatdan) [inline libkpod/remove.go]: errors.Wrapf(err, ...)

File: libkpod/remove.go [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+			}
+		}
+	}
+	if err := c.runtime.DeleteContainer(ctr); err != nil {
+		return "", fmt.Errorf("failed to delete container %s: %v", ctrID, err)
```
Comment (rhatdan) [inline libkpod/remove.go]: errors.Wrapf(err, ...)

File: libkpod/remove.go [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	c.RemoveContainer(ctr)
```
Comment (rhatdan) [inline libkpod/remove.go]: Does this return an error?

File: libkpod/remove.go [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+	if err := c.storageRuntimeServer.DeleteContainer(ctrID); err != nil {
+		return "", fmt.Errorf("failed to delete storage for container %s: %v", ctrID, err)
+	}
+
+	c.ReleaseContainerName(ctr.Name())
```
Comment (rhatdan) [inline libkpod/remove.go]: Does this return an error?

File: libkpod/remove.go [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+		}
+	}
+	if err := c.runtime.DeleteContainer(ctr); err != nil {
+		return "", fmt.Errorf("failed to delete container %s: %v", ctrID, err)
+	}
+
+	c.RemoveContainer(ctr)
+
+	if err := c.storageRuntimeServer.DeleteContainer(ctrID); err != nil {
+		return "", fmt.Errorf("failed to delete storage for container %s: %v", ctrID, err)
```
Comment (rhatdan) [inline libkpod/remove.go]: errors.Wrapf(err, ...)

File: libkpod/stop.go [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+		return "", err
+	}
+
+	if err := c.runtime.UpdateStatus(ctr); err != nil {
+		return "", err
+	}
+	cStatus := c.runtime.ContainerStatus(ctr)
+	if cStatus.Status != oci.ContainerStateStopped {
+		if err := c.runtime.StopContainer(ctr, timeout); err != nil {
+			return "", fmt.Errorf("failed to stop container %s: %v", ctr.ID(), err)
```
Comment (rhatdan) [inline libkpod/stop.go]: errors.Wrapf(err, ...)

File: libkpod/stop.go [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/kubernetes-incubator/cri-o/oci"
+	"github.com/sirupsen/logrus"
+	pb "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+)
+
+// Stop stops a running container with a grace period (i.e., timeout).
+func (c *ContainerServer) ContainerStop(container string, timeout int64) (string, error) {
+	ctr, err := c.LookupContainer(container)
+	if err != nil {
+		return "", err
```
Comment (rhatdan) [inline libkpod/stop.go]: errors.Wrapf(err, Failed to stop conatainer...)

File: libkpod/stop.go [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+	}
+
+	if err := c.runtime.UpdateStatus(ctr); err != nil {
+		return "", err
```
Comment (rhatdan) [inline libkpod/stop.go]: Does this return a decent error message?

File: libkpod/stop.go [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	cStatus := c.runtime.ContainerStatus(ctr)
+	if cStatus.Status != oci.ContainerStateStopped {
+		if err := c.runtime.StopContainer(ctr, timeout); err != nil {
+			return "", fmt.Errorf("failed to stop container %s: %v", ctr.ID(), err)
+		}
+		if err := c.storageRuntimeServer.StopContainer(ctr.ID()); err != nil {
+			return "", fmt.Errorf("failed to unmount container %s: %v", ctr.ID(), err)
```
Comment (rhatdan) [inline libkpod/stop.go]: errors.Wrapf(err, ...)

File: libkpod/stop.go [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+		}
+	}
+
+	c.ContainerStateToDisk(ctr)
```
Comment (rhatdan) [inline libkpod/stop.go]: Does this return errors?

File: server/container_remove.go [hunk 24]
```diff
@@ -12,42 +9,11 @@ import (
 // RemoveContainer removes the container. If the container is running, the container
 // should be force removed.
 func (s *Server) RemoveContainer(ctx context.Context, req *pb.RemoveContainerRequest) (*pb.RemoveContainerResponse, error) {
-	logrus.Debugf("RemoveContainerRequest %+v", req)
-	c, err := s.GetContainerFromRequest(req.ContainerId)
+	_, err := s.ContainerServer.Remove(req.ContainerId, true)
```
Comment (rhatdan) [inline server/container_remove.go]: RemoveContainer?

File: server/container_stop.go [hunk 25]
```diff
@@ -1,37 +1,18 @@
...
 	"github.com/sirupsen/logrus"
 	"golang.org/x/net/context"
 	pb "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
 )
 
 // StopContainer stops a running container with a grace period (i.e., timeout).
 func (s *Server) StopContainer(ctx context.Context, req *pb.StopContainerRequest) (*pb.StopContainerResponse, error) {
-	logrus.Debugf("StopContainerRequest %+v", req)
-	c, err := s.GetContainerFromRequest(req.ContainerId)
+	_, err := s.ContainerServer.Stop(req.ContainerId, req.Timeout)
```
Comment (rhatdan) [inline server/container_stop.go]: ContainerStop?

File: test/kpod_rm.bats [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+    cleanup_test
+}
+
+@test "remove a container" {
+    start_crio
+    run crioctl pod run --config "$TESTDATA"/sandbox_config.json
+    echo "$output"
+    [ "$status" -eq 0 ]
+    pod_id="$output"
+	run crioctl ctr create --config "$TESTDATA"/container_config.json --pod "$pod_id"
```
Comment (rhatdan) [inline test/kpod_rm.bats]: Indent problem

File: libkpod/remove.go (hunk 16 above)
Comment (mheon) [inline libkpod/remove.go]: Not yet, but it should. Working on that code now.

File: libkpod/remove.go (hunk 17 above)
Comment (mheon) [inline libkpod/remove.go]: Nope, that operation cannot fail

File: libkpod/remove.go (hunk 16 above)
Comment (14rcole) [inline libkpod/remove.go]: No

File: libkpod/remove.go (hunk 17 above)
Comment (14rcole) [inline libkpod/remove.go]: No it doesn't

File: server/container_remove.go (hunk 24 above)
Comment (14rcole) [inline server/container_remove.go]: ContainerRemove().  RemoveContainer() already existed so I couldn't use that.  I do need to change it though

File: server/container_stop.go (hunk 25 above)
Comment (14rcole) [inline server/container_stop.go]: Yes

File: completions/bash/kpod (hunk 10 above)
Comment (14rcole) [inline completions/bash/kpod]: Not yet

**Comment by 14rcole** [comment]
@mheon Any idea why runtime.UpdateStatus() is reporting stopped for a container that I had just started with crio?

**Comment by mheon** [comment]
Could be the process in the container immediately exited

**Comment by 14rcole** [comment]
You're right, it does seem to be exiting immediately.  I'll try to figure out why

**Comment by 14rcole** [comment]
@mrunalp @runcom PTAL

**Comment by TomSweeneyRedHat** [comment]
@14rcole looks like a rebase is needed

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
[retest]

**Comment by 14rcole** [comment]
@mrunalp All the tests that are failing are doing so in cleanup, and I can't tell why.  Do you have any thoughts?  Or is this connected to other issues in Jenkins?

File: cmd/kpod/rm.go [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+
+	config, err := getConfig(c)
+	if err != nil {
+		return errors.Wrapf(err, "could not get config")
+	}
+	server, err := libkpod.New(config)
+	if err != nil {
+		return errors.Wrapf(err, "could not get container server")
+	}
+	err = server.Update()
```
Comment (nalind) [inline cmd/kpod/rm.go]: Needs a `defer server.Shutdown()` here.

**Comment by nalind** [comment]
At startup, the storage library's overlay driver bind mounts the storage root over itself to set it as private, and we're not shutting it down, so it doesn't get unmounted, and the filesystem remains busy.  We aren't seeing this in Travis because we have to use the vfs driver there, and the vfs driver doesn't bind mount the storage root over itself to mark that part of the filesystem as private.  #778 should address part of this.

File: cmd/kpod/rm.go (hunk 27 above)
Comment (nalind) [inline cmd/kpod/rm.go]: (Before we call `server.Update()`.)

**Comment by rhatdan** [comment]
@mrunalp can you see why https://github.com/kubernetes-incubator/cri-o/pull/778 E2E tests are not running?

File: docs/kpod.1.md [hunk 28]
```diff
@@ -35,8 +35,8 @@ has the capability to debug pods/images created by crio.
 **--runroot**=**value**
    Path to the 'run directory' where all state information is stored
 
-**--runtime**=**value**
-    Path to the OCI-compatible binary used to run containers
+**--runtime**="**value**
+    Path to the OCI oci-compatible binary used to run containers
```
Comment (nalind) [inline docs/kpod.1.md]: Were these left over from a rebase?

Comment (14rcole) [inline docs/kpod.1.md]: Whoops, yes.  Fixed now

**Comment by 14rcole** [comment]
retest

**Comment by rhatdan** [comment]
Replaced this pull request with https://github.com/kubernetes-incubator/cri-o/pull/783
Ryan has gone back to school.

**Pull request closed without merging**